oauth-server/src/main/java/io/strimzi/kafka/oauth/server/JaasServerOauthValidatorCallbackHandler.java

```diff
@@ -0,0 +1,222 @@
+import java.util.Set;
+
+import static io.strimzi.kafka.oauth.common.JSONUtil.getClaimFromJWT;
+import static io.strimzi.kafka.oauth.common.LogUtil.mask;
+
+public class JaasServerOauthValidatorCallbackHandler implements AuthenticateCallbackHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(JaasServerOauthValidatorCallbackHandler.class);
+
+    private TokenValidator validator;
+
+    private ServerConfig config;
+
+    private String usernameClaim;
+
+    @Override
+    public void configure(Map<String, ?> configs, String saslMechanism, List<AppConfigurationEntry> jaasConfigEntries) {
+
+        if (!OAuthBearerLoginModule.OAUTHBEARER_MECHANISM.equals(saslMechanism))    {
+            throw new IllegalArgumentException(String.format("Unexpected SASL mechanism: %s", saslMechanism));
+        }
+
+        if (jaasConfigEntries.size() != 1) {
+            throw new IllegalArgumentException("Exactly one jaasConfigEntry expected (size: " + jaasConfigEntries.size());
+        }
+
+        for (AppConfigurationEntry e: jaasConfigEntries) {
+            Properties p = new Properties();
+            p.putAll(e.getOptions());
+            config = new ServerConfig(p);
+            break;
+        }
+
+        validateConfig();
+
+        SSLSocketFactory socketFactory = ConfigUtil.createSSLFactory(config);
+        HostnameVerifier verifier = ConfigUtil.createHostnameVerifier(config);
+
+        String jwksUri = config.getValue(ServerConfig.OAUTH_JWKS_ENDPOINT_URI);
+
+        if (jwksUri != null) {
+            validator = new JWTSignatureValidator(
+                    config.getValue(ServerConfig.OAUTH_JWKS_ENDPOINT_URI),
+                    socketFactory,
+                    verifier,
+                    config.getValue(ServerConfig.OAUTH_VALID_ISSUER_URI),
+                    config.getValueAsInt(ServerConfig.OAUTH_JWKS_REFRESH_SECONDS, 300),
+                    config.getValueAsInt(ServerConfig.OAUTH_JWKS_EXPIRY_SECONDS, 360),
+                    config.getValueAsBoolean(ServerConfig.OAUTH_VALIDATE_COMMON_CHECKS, true),
+                    config.getValue(ServerConfig.OAUTH_VALIDATE_AUDIENCE)
+            );
+        } else {
+            validator = new OAuthIntrospectionValidator(
+                    config.getValue(ServerConfig.OAUTH_INTROSPECTION_ENDPOINT_URI),
+                    socketFactory,
+                    verifier,
+                    config.getValue(ServerConfig.OAUTH_VALID_ISSUER_URI),
+                    config.getValue(Config.OAUTH_CLIENT_ID),
+                    config.getValue(Config.OAUTH_CLIENT_SECRET),
+                    config.getValueAsBoolean(ServerConfig.OAUTH_VALIDATE_COMMON_CHECKS, true),
+                    config.getValue(ServerConfig.OAUTH_VALIDATE_AUDIENCE)
+            );
+        }
+
+        usernameClaim = config.getValue(Config.OAUTH_USERNAME_CLAIM, "sub");
+        if ("sub".equals(usernameClaim)) {
+            usernameClaim = null;
+        }
```

Comment:
Why does "sub" get mapped to null?

Reply:
"sub" is the default claim used for username, meaning that in this case username mapping is not required.
usernameClaim != null means more complex code path further down - to map the username.